lib/lz4.c

```diff
@@ -692,16 +747,27 @@ LZ4_FORCE_INLINE int LZ4_compress_generic(
 
         /* Test next position */
         match = LZ4_getPosition(ip, cctx->hashTable, tableType, base);
-        if (dict==usingExtDict) {
+        if (dictDirective == usingExtDictCtx) {
             if (match < (const BYTE*)source) {
+                /* there was no match, try the dictionary */
+                /* TODO: use precalc-ed hash? */
+                match = LZ4_getPosition(ip, dictCtx->hashTable, byU32, dictBase);
                 refDelta = dictDelta;
-                lowLimit = dictionary;
+                lowLimit = dictLowLimit;
+            } else {
+                refDelta = 0;
+                lowLimit = (const BYTE*)source;
+            }
+        } else if (dictDirective==usingExtDict) {
+            if (match < (const BYTE*)source) {
+                refDelta = dictDelta;
+                lowLimit = dictLowLimit;
             } else {
                 refDelta = 0;
                 lowLimit = (const BYTE*)source;
         }   }
         LZ4_putPosition(ip, cctx->hashTable, tableType, base);
-        if ( ((dictIssue==dictSmall) ? (match>=lowRefLimit) : 1)
+        if ( match >= lowRefLimit
```

Comment:
I guess this test is now run at each loop,
while previously, it was only active when `(dictIssue == dictSmall)`.
This can make a difference.

Reply:
Yes, unfortunately, by not resetting the table as frequently (and by starting the context with non-zero `currentOffset`), there are pretty limited situations in which this optimization can be restored. The branch `https://github.com/felixhandte/lz4/tree/ref-dict-table-test-2` tries to restore the check, but benchmarking didn't show an improvement.